plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+        what fields are felt by the particles. Options are:
+
+        * 'nearest neighbor': Particles are assigned the fields on
+            the grid vertex closest to them.
+        * 'volume averaged' : The fields experienced by a particle are a
+            volume-average of the eight grid points surrounding them.
+
+        The default is 'volume averaged'.
+
+    req_quantities : `list` of str, optional
+        A list of quantity keys required to be specified on the Grid object.
+        The base particle pushing simulation requires the quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        If any of these quantities are missing, a warning will be given and that
+        quantity will be assumed to be zero everywhere. This keyword is for specifying quantities in
+        addition to these six. The default is None.
+
+    verbose : bool, optional
+        If true, updates on the status of the program will be printed
+        into the standard output while running. The default is True.
+    """
+
+    def __init__(
+        self,
+        grids: Union[AbstractGrid, Iterable[AbstractGrid]],
+        termination_condition: AbstractTerminationCondition = None,
+        save_routine: Optional[AbstractSaveRoutine] = None,
+        dt=None,
+        dt_range=None,
+        field_weighting="volume averaged",
+        req_quantities=None,
+        verbose=True,
+    ):
+        # self.grid is the grid object
+        if isinstance(grids, AbstractGrid):
+            self.grids = [
+                grids,
+            ]
+        elif isinstance(grids, collections.abc.Iterable):
+            self.grids = grids
+        elif isinstance(grids, BasePlasma):
+            raise TypeError(
+                "It appears you may be trying to access an older version of the ParticleTracker class."
+                "This class has been deprecated."
+                "Please revert to PlasmaPy version 2023.5.1 to use this version of ParticleTracker."
+            )
+        else:
+            raise TypeError("Type of argument `grids` not recognized.")
+
+        # self.grid_arr is the grid positions in si units. This is created here
+        # so that it isn't continuously called later
+        self.grids_arr = [grid.grid.to(u.m).value for grid in self.grids]
+
+        self.verbose = verbose
+
+        # This flag records whether the simulation has been run
+        self._has_run = False
+
+        self._is_adaptive_time_step = False
+        self._is_synchronized_time_step = False
+
+        # By default, the gyration of a particle is divided into twelve steps
+        self._steps_per_gyroperiod = 12
+
+        self._Courant_parameter = 0.5
+
+        # *********************************************************************
+        # Validate required fields
+        # *********************************************************************
+
+        base_req_quantities = ["E_x", "E_y", "E_z", "B_x", "B_y", "B_z"]
```

Comment:
Thinking about how we should handle this for cases where we might add more quantities to this list in response to other keywords. I'd suggest 

1) Make this list a private attribute of the class, define right at the beginning of init 
2) If other keywords are turned on (e.g. stopping soon) we can then append to this list
3) Then have this validate structure at the very end of init. To avoid init being a long method, maybe break this into a separate method `_validate_grids()`

Reply:
(use set)

Reply:
And it should be a set! Good call